hedera-node/hedera-app/src/main/java/com/hedera/node/app/records/BlockRecordManagerImpl.java

```diff
@@ -0,0 +1,482 @@
+ *         <li>Manages Record State</li>
+ *     </ul>
+ *     </li>
+ *     <li><b>StreamFileProducer ({@link StreamFileProducerBase} and {@link StreamFileProducerConcurrent})</b> - Responsible for managing
+ *     production record and sidecar streams in a multi-threaded efficient way.</li>
+ *     <li><b>File Writers({@link RecordFileWriter}, {@link SidecarFileWriter} and {@link SignatureFileWriter})</b> - Responsible for writing the actual files
+ *     in the correct format.</li>
+ * </ul>
+ * <p>
+ *     NOTE: It expects all methods to be called on handle transaction thread.
+ * </p>
+ * <p>
+ *     NOTE: Any state that needs to persist across restarts is stored in the {@link BlockInfo} state.
+ * </p>
+ */
+@SuppressWarnings("DuplicatedCode")
+@Singleton
+public final class BlockRecordManagerImpl implements BlockRecordManager {
+    private static final Logger log = LogManager.getLogger(BlockRecordManagerImpl.class);
+    /** The hash size in bytes, normally 48 for SHA384 */
+    private static final int HASH_SIZE = DigestType.SHA_384.digestLength();
+
+    /** Number of blocks to keep multiplied by hash size */
+    private final int numBlockHashesToKeepBytes;
+    /** The number of milliseconds of consensus time in a block period, from configuration */
+    private final long blockPeriodInMilliSeconds;
+    /** The configuration provider */
+    private final ConfigProvider configProvider;
+    /** The stream file producer we are using */
+    private final StreamFileProducerBase streamFileProducer;
+    /** The working state accessor, for accessing state adhoc */
+    private final WorkingStateAccessor workingStateAccessor;
+    /** Ture when we have completed event recovery */
+    private boolean eventRecoveryCompleted = false;
+    /** The number of the current block. "provisional" because the block is not yet complete. */
+    private long provisionalCurrentBlockNumber;
+    /** The consensus time of the first transaction in the current block. "provisional" because the block is not yet complete. */
+    private Instant provisionalCurrentBlockFirstTransactionTime = null;
+
+    /**
+     * Construct BlockRecordManager
+     *
+     * @param configProvider The configuration provider
+     * @param nodeInfo The node info
+     * @param signer The signer for signing files
+     * @param fileSystem the file system to use, needed for testing to be able to use a non-standard file
+     *                   system. If null default is used.
+     * @param workingStateAccessor The working state accessor, for accessing state adhoc
+     */
+    @Inject
+    public BlockRecordManagerImpl(
+            @NonNull final ConfigProvider configProvider,
+            @NonNull final NodeInfo nodeInfo,
+            @NonNull final Signer signer,
+            @Nullable final FileSystem fileSystem,
+            @NonNull final WorkingStateAccessor workingStateAccessor) {
+        this.configProvider = requireNonNull(configProvider);
```

Comment:
We should try to avoid using `ConfigProvider` when possible. Can we pass the current configuration as a parameter in `startUserTransaction()` etc.?

Why is a `WorkingStateAccessor` needed?

Reply:
`workingStateAccessor` is needed for methods like `lastBlockNo()` which is a API BlockRecordManager provide to other services (Util & Smart Contract). Those getter methods need to read the state. This is how Richard recommended. Is there a better way for them to get the right state to read from?

Reply:
For the config question I get two properties from config during construction. This means they are not dynamic properties. I was debating this with Richard. I find it hard to face the performance cost of checking all these properties on every transaction when they will almost never change. Also I was not clear how we handle determinism in config property changes? changing these in non-deterministic way would cause a ISS.

Reply:
The issue with `workingStateAccessor` is similar to the one mentioned above. The returned state will not be aware of written but not committed changes. This is probably ok for the current code. But it introduces a risk because one needs to be aware of this when making future changes.

Property changes are done via transactions and, therefore, deterministic. (I am not 100% sure when exactly the change should happen, but it will be implemented deterministically.)

I am also not happy with the way we handle configs right now. It would be nice if we could distinguish static and dynamic properties. Another option I am considering is a mechanism that automatically generates new facilities when the configuration changes. Each facility could then assume that its configuration is immutable. But it felt like premature optimization to do this right now.